libs/appstore/src/appstore/appstore-app-client.ts

```diff
@@ -140,6 +147,235 @@ export class AppstoreAppClient {
+      },
+      dnaZomeFunction: {
+        dna: appVersion.apphub_hrl.dna,
+        zome: 'apphub_csr',
+        function: 'get_webapp_package_version_entry', // We just pick one of the functions for the sake of simplicity and assume that all other functions are callable as well by the same host
+      },
+      pingTimeout: 4000,
+    });
+  }
+
+  /**
+   *
+   * @param appVersion
+   * @param webappPackageVersion Can be provided if the WebAppPackageVersionEntry has already been fetched
+   * otherwise in order to omit redundant remote calls
+   * @returns
+   */
+  async fetchHappBytes(
+    appVersion: AppVersionEntry,
+    webappPackageVersion?: WebAppPackageVersionEntry,
+  ): Promise<Uint8Array> {
+    // For simplicity make all calls with one host. If that proves to not work well, split into
+    // separate calls to different hosts
+    return this.portalZomeClient.tryWithHosts<Uint8Array>({
+      fn: async (host) => {
+        // 1. get WebappPackageVersion and verify its hash - only necessary if the entry is not already
+        // provided as a function argument.
+        if (!webappPackageVersion) {
+          webappPackageVersion =
+            await this.portalZomeClient.customRemoteCall<WebAppPackageVersionEntry>({
+              host,
+              call: {
+                dna: appVersion.apphub_hrl.dna,
+                zome: 'apphub_csr',
+                function: 'get_webapp_package_version_entry',
+                payload: appVersion.apphub_hrl.target,
+              },
+            });
+
+          console.log('@fetchHappBytes Got WebAppPackageVersionEntry: ', webappPackageVersion);
+        }
+        // validate hash of received entry
+        const webappPackageVersionEntryHash =
+          await this.appstoreZomeClient.hashWebappPackageVersionEntry(webappPackageVersion);
+        if (
+          encodeHashToBase64(webappPackageVersionEntryHash) !==
+          encodeHashToBase64(appVersion.apphub_hrl_hash)
+        ) {
+          throw new Error(
+            `Hash of WebappPackageVersionEntry does not match expected hash. Got ${encodeHashToBase64(webappPackageVersionEntryHash)}, expected ${encodeHashToBase64(appVersion.apphub_hrl_hash)}`,
+          );
+        }
+
+        // 2. Get the WebappEntry to figure out the entry hash of the UI
+        const webappEntryEntity = await this.portalZomeClient.customRemoteCall<Entity<WebAppEntry>>(
+          {
+            host,
+            call: {
+              dna: appVersion.apphub_hrl.dna,
+              zome: 'apphub_csr',
+              function: 'get_webapp_entry',
+              payload: webappPackageVersion.webapp,
+            },
+          },
+        );
+        // validate hash of received entry
+        const webappEntryHash = await this.appstoreZomeClient.hashWebappEntry(
+          webappEntryEntity.content,
+        );
+        if (webappEntryHash.toString() !== webappPackageVersion.webapp.toString()) {
```

Comment:
Any reason why you choose to compare hashes this way instead of comparing their encoded base64 strings like I've seen done above?

Reply:
Thanks! I should in fact compare the base64 encoded ones as `.toString()` has proven unreliable for the purpose of comparing for some reason, i.e. there have been cases where a hash would not be equal when comparing with `.toString()` but they were equal when comparing the base64 encoded strings.

Reply:
Pushed a commit with this fixed.